### PR TITLE
Raise PackageVersionLookupNotSupportedError

### LINKED ISSUE #1727: Public typed error for unsupported package version lookup
**Need:** API consumers calling `getPackageAvailableVersions` cannot tell the difference between a package manager that genuinely does not support version lookup (an unsupported *capability*) and an operational failure (a failed command, network error, malformed output, or parsing miss). Today both cases collapse to `undefined`, so consumers cannot decide whether to fall back to manual version entry or surface a real error. The reference work in #1717 introduced this distinction only as an *internal* error; it should be part of the **public** API contract so downstream extensions can react to it reliably across bundle boundaries.

**Proposal:**

- Add a public, exported `PackageVersionLookupNotSupportedError` (with a stable `code` discriminator) plus an `isPackageVersionLookupNotSupportedError` type guard that works across extension bundle boundaries (where `instanceof` is unreliable).
- Update `PythonPackageGetterApi.getPackageAvailableVersions` so that:
  - Unsupported capability (default/missing manager, Poetry, Pip older than 21.2) rejects with the typed error.
  - Operational failures (Pip/Conda command, network, malformed-output, parsing) propagate as their original errors instead of becoming `undefined`.
- Bump the public API package version and document the change in the API changelog.
- Update consumers so that only the typed unsupported error triggers the manual version-entry fallback; all other errors propagate for normal handling.

**Acceptance Criteria:**

- `PackageVersionLookupNotSupportedError` and `isPackageVersionLookupNotSupportedError` are exported from the public API (`src/api.ts`) and the `@vscode/python-environments` package.
- The type guard recognizes the error via its stable `code` even when the error object was created by a different bundle copy of the class.
- Default adapter (no/missing manager), Poetry, and Pip < 21.2 reject with `PackageVersionLookupNotSupportedError`.
- Pip/Conda command, network, malformed-output, and parsing failures reject with their original errors (not `undefined`).
- Interpreter version strings with release-level/serial suffixes (e.g. `3.13.14.final.0`) resolve correctly for Pip's `--python-version` flag.
- The "Manage Package Version" UI falls back to manual entry only on the typed unsupported error; other errors propagate.
- The public API package version is bumped and the API changelog documents the new error, type guard, and refined contract.
- Focused unit tests cover the default adapter, Poetry, old Pip, supported-Pip command failures, Conda failures, cross-bundle recognition, and the `3.13.14.final.0` version string.

Follow-up to #1717 (which added the internal-only error and cross-platform integration coverage).

### PR DESCRIPTION
**Pull request by edvilme**
Fixes #1727

This pull request adds explicit package-version lookup errors without breaking existing callers.

## API behavior

- Adds `PackageVersionLookupNotSupportedError` with the stable code `PackageVersionLookupNotSupported`.
- Adds `isPackageVersionLookupNotSupportedError()` for reliable detection across extension bundle boundaries.
- Adds an optional `errorMode` parameter to `PythonPackageGetterApi.getPackageAvailableVersions()`:
  - Omitted or `legacy`: preserves the existing behavior and resolves `undefined` for unsupported lookups and operational failures.
  - `throw`: rejects with `PackageVersionLookupNotSupportedError` when lookup is unsupported and propagates operational failures unchanged.
- Provides overloads so throw mode returns `Promise<Pep440Version[]>`, while legacy mode returns `Promise<Pep440Version[] | undefined>`.

## Implementations

- Pip reports pip versions older than 21.2 as unsupported and propagates command or parsing failures in throw mode.
- Conda propagates command and parsing failures in throw mode.
- Poetry reports version lookup as unsupported.
- Missing and third-party package-manager implementations are adapted according to the requested error mode.
- The extension-owned package version picker opts into throw mode, falling back to manual entry only for the typed unsupported error.
- Python interpreter versions such as `3.13.14.final.0` are normalized for pip lookup.

## Package and tests

- Bumps `@vscode/python-environments` to 1.3.0 and updates the API changelog and consumer type checks.
- Adds focused coverage verifying the default legacy behavior and the opt-in throw mode.

### CODE REVIEW

**Comment by heejaechang** [comment]
🔒 Automated review in progress — @heejaechang is auto-reviewing this PR.

File: api/package.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
     "name": "@vscode/python-environments",
     "description": "An API facade for the Python Environments extension in VS Code",
-    "version": "1.2.0",
+    "version": "1.3.0",
```
Comment (heejaechang) [inline api/package.json:4]: > **Issue** · Please address or respond

Changing `getPackageAvailableVersions` from resolving `undefined` to rejecting—and removing `undefined` from its public return type—is a breaking API change, but this publishes it as `1.3.0` despite the package's stated SemVer policy. Use a major version or provide a compatibility-preserving migration path.

[verified]

File: src/api.ts [hunk 2]
```diff
@@ -1139,18 +1200,24 @@ export interface PythonPackageGetterApi {
...
      *
-     * Support depends on the package manager backing the environment. Managers that do
-     * not implement version lookup resolve to `undefined`.
+     * The returned promise distinguishes an unsupported capability from an operational failure:
+     * - It resolves to a non-empty array of {@link Pep440Version} objects when versions are found.
+     * - It rejects with a {@link PackageVersionLookupNotSupportedError} when the environment's
+     *   package manager does not support version lookup (for example, the default/missing manager,
+     *   Poetry, or a Pip older than 21.2). Use {@link isPackageVersionLookupNotSupportedError} to
+     *   detect this reliably across extension bundle boundaries and fall back to manual entry.
+     * - It rejects with the original error for any other failure (command, network, or
```
Comment (heejaechang) [inline src/api.ts]: > **Warning** · Non-blocking recommendation

The API promises a non-empty success result, but Pip accepts `[]`, Conda can produce `[]` after filtering, and the adapter does not reject empty arrays. Either enforce non-empty results at a shared boundary and test them, or remove this guarantee from the API and changelog.

[verified]

File: api/test/consumer.ts [hunk 3]
```diff
@@ -4,20 +4,39 @@ import type {
...
+const availableVersions: Promise<Pep440Version[]> = api.getPackageAvailableVersions(environment, 'example');
+
+// The unsupported-capability error is part of the public contract: it is constructible, extends
+// Error, and exposes a stable string-literal `code` discriminator.
+const lookupError = new PackageVersionLookupNotSupportedError('unsupported');
+const lookupErrorIsError: Error = lookupError;
+const lookupErrorCodeIsExact: Equal<typeof lookupError.code, 'PackageVersionLookupNotSupported'> = true;
+
+// The type guard narrows unknown values via the stable discriminator (bundle-boundary safe).
+declare const maybeError: unknown;
```
Comment (heejaechang) [inline api/test/consumer.ts:43]: > **Warning** · Non-blocking recommendation

These assertions verify only the exported types. Add focused runtime tests for old Pip, empty/all-invalid output, command failures, missing executables, legacy `undefined`, and cross-bundle-shaped errors so the new classification behavior is covered.

[verified]

**Comment by heejaechang** [comment]
**Verification:** The relevant tests could not be fully run in the isolated environment; this review is not fully verified.

File: api/package.json (hunk 1 above)
Comment (heejaechang) [inline api/package.json:4]: > **Issue** · Please address or respond

The updated method can now reject in cases that previously resolved to `undefined`, which is a breaking runtime contract change for existing consumers. Publish this under a major version, or provide a compatibility-safe API surface.

File: src/api.ts [hunk 4]
```diff
@@ -1139,18 +1200,24 @@ export interface PythonPackageGetterApi {
     /**
      * Get the list of available versions for a package, newest first.
      *
-     * Support depends on the package manager backing the environment. Managers that do
-     * not implement version lookup resolve to `undefined`.
+     * The returned promise distinguishes an unsupported capability from an operational failure:
+     * - It resolves to a non-empty array of {@link Pep440Version} objects when versions are found.
```
Comment (heejaechang) [inline src/api.ts]: > **Issue** · Please address or respond

The documented non-empty result guarantee is not enforced: Pip, Conda, and the internal adapter can still return an empty array. Reject empty/all-invalid results at a shared boundary and test it, or remove the non-empty guarantee from the API documentation and changelog.

**Comment by heejaechang** [comment]
**Verification:** The relevant tests could not be fully run in the isolated environment; this review is not fully verified.

**Comment by heejaechang** [comment]
**Verification:** The relevant tests could not be fully run in the isolated environment; this review is not fully verified.

File: src/managers/builtin/pipPackageManager.ts [hunk 5]
```diff
@@ -244,6 +279,33 @@ export class PipPackageManager implements PackageManager, Disposable {
     }
 }
 
+/**
+ * Ensures a parse step produced versions, converting a parsing miss into a propagating
+ * operational error instead of silently returning `undefined`.
+ */
```
Comment (heejaechang) [inline src/managers/builtin/pipPackageManager.ts:285]: > **Issue** · Please address or respond

`requireParsedVersions` only rejects `undefined`, so an empty array is treated as a successful lookup despite the new non-empty-result contract. Conda can likewise return an empty array after filtering, and the internal adapter forwards it. Reject empty results as operational failures at the shared boundary (preserving legacy `undefined` behavior), and cover empty/all-invalid results in the adapter and manager tests.

**Comment by heejaechang** [comment]
**Verification:** The relevant tests could not be fully run in the isolated environment; this review is not fully verified.

File: src/internal.api.ts [hunk 6]
```diff
@@ -400,10 +402,42 @@ export class InternalPackageManager implements PackageManager {
...
-            : Promise.resolve(undefined);
+        const shouldThrow = options?.errorMode === 'throw';
+        try {
+            if (!this.manager.getPackageAvailableVersions) {
+                throw new PackageVersionLookupNotSupportedError(
+                    `Package version lookup is not supported by package manager: ${this.id}`,
+                );
+            }
+            const versions = await this.manager.getPackageAvailableVersions(environment, packageName);
+            if (versions === undefined && shouldThrow) {
```
Comment (heejaechang) [inline src/internal.api.ts:429]: > **Issue** · Please address or respond

This compatibility boundary forwards an empty array as a successful result, but the public contract promises a non-empty array. Pip parsing and Conda's invalid-entry filtering can both produce `[]`; treat that as an operational failure in `throw` mode (and `undefined` in legacy mode), and cover the empty-result cases.

**Comment by heejaechang** [comment]
**Verification:** The relevant tests could not be fully run in the isolated environment; this review is not fully verified.

**Comment by eleanorjboyd** [comment]
heejae's review lgtm - I agree with those items and once those fixes are in I can approve it

File: src/test/internalPackageManager.versionLookup.unit.test.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        const manager = new InternalPackageManager('test:manager', {} as unknown as PackageManager);
+        assert.strictEqual(await manager.getPackageAvailableVersions(environment, 'requests'), undefined);
+        await assert.rejects(
+            () => manager.getPackageAvailableVersions(environment, 'requests', { errorMode: 'throw' }),
+            (error: unknown) => isPackageVersionLookupNotSupportedError(error),
+        );
+    });
+
+    test('handles operational errors according to the requested error mode', async () => {
+        const operational = new Error('network down');
```
Comment (heejaechang) [inline src/test/internalPackageManager.versionLookup.unit.test.ts]: > **Warning** · Non-blocking recommendation

Add coverage for empty/all-invalid delegated results and focused built-in-manager cases for old Pip, missing executables, and command failures. These cases are needed to verify that legacy mode normalizes failures while throw mode preserves operational errors and reports only unsupported capability with the typed error.

**Comment by heejaechang** [comment]
**Verification:** The relevant tests could not be fully run in the isolated environment; this review is not fully verified.

File: src/api.ts (hunk 4 above)
Comment (edvilme) [reply src/api.ts]: This outdated thread is superseded by the current empty-result discussions on `pipPackageManager.ts` and `internal.api.ts`; those threads remain open.

File: src/test/internalPackageManager.versionLookup.unit.test.ts (hunk 7 above)
Comment (edvilme) [reply src/test/internalPackageManager.versionLookup.unit.test.ts]: The test suite was simplified to cover the two externally observable parameter behaviors directly: omitted mode resolves `undefined`, and throw mode rejects. This outdated thread is superseded by the focused compatibility-boundary tests.

File: api/test/consumer.ts (hunk 3 above)
Comment (edvilme) [reply api/test/consumer.ts:43]: Added focused runtime coverage at the compatibility boundary: one test calls the method without the parameter and verifies legacy `undefined`, and one passes `{ errorMode: 'throw' }` and verifies rejection. We intentionally kept this coverage at the public behavior boundary rather than duplicating manager implementation tests.

File: api/package.json (hunk 1 above)
Comment (edvilme) [reply api/package.json:4]: Addressed with a compatibility-safe API surface: omitted calls preserve the legacy `undefined` behavior and return type, while rejection behavior is available only through the explicit `{ errorMode: 'throw' }` option.

File: src/api.ts (hunk 2 above)
Comment (edvilme) [reply src/api.ts]: This outdated thread is superseded by the current empty-result discussions on `pipPackageManager.ts` and `internal.api.ts`; those threads remain open.

File: api/package.json (hunk 1 above)
Comment (edvilme) [reply api/package.json:4]: Addressed by adding the optional `errorMode` parameter. Calls that omit it retain the existing `Promise<Pep440Version[] | undefined>` contract and normalize failures to `undefined`; only callers that explicitly pass `{ errorMode: 'throw' }` opt into rejection behavior.

**Comment by heejaechang** [comment]
**Verification:** The relevant tests could not be fully run in the isolated environment; this review is not fully verified.

File: src/managers/builtin/pipPackageManager.ts (hunk 5 above)
Comment (edvilme) [reply src/managers/builtin/pipPackageManager.ts:285]: Addressed by defining an empty array as a valid successful lookup result. The non-empty guarantee was removed from the Pip and package-manager documentation; the existing UI already falls back to manual entry when the returned array is empty.

File: src/internal.api.ts (hunk 6 above)
Comment (edvilme) [reply src/internal.api.ts:429]: Addressed by allowing empty arrays as valid successful results and removing the non-empty guarantee from the API documentation. The compatibility adapter therefore intentionally forwards `[]` in both modes.

**Comment by heejaechang** [comment]
**Verification:** The relevant tests could not be fully run in the isolated environment; this review is not fully verified.

**Review by heejaechang — APPROVED** [review]
Approved via Review Center.

**Review by StellaHuang95 — APPROVED** [review]
Approved via Review Center.

**Review by StellaHuang95 — APPROVED** [review]
Approved via Review Center.

**Review by StellaHuang95 — APPROVED** [review]
Approved via Review Center.

**Review by StellaHuang95 — APPROVED** [review]
Approved via Review Center.
